### ISSUE TITLE
Add MaxLength to TextBox

### ISSUE THREAD

**Issue by rdiazconcha** [Microsoft/xaml-standard#18]

**Comment by MovGP0** [comment]
Can we replace `MaxLength` and `MinLength` with validators? This would be much more flexible, since you can simply add custom validators as needed. 

```xaml
<TextBox x:Name="firstNameTextBox">
    <TextBox.Validators>
        <NotEmptyValidator                   Message="..." Enabled="{Binding ...}" />
        <MinLengthValidator MinLength="5"    Message="..." Enabled="{Binding ...}" />
        <MaxLengthValidator MaxLength="30"   Message="..." Enabled="{Binding ...}" />
        <RegexValidator Pattern="[a-zA-Z ]+" Message="..." Enabled="{Binding ...}" />
    </TextBox.Validators>
</TextBox>
```

**Comment by birbilis** [comment]
ASP.net WebForms had nice validators and summary etc.

**Comment by insinfo** [comment]
@rdiazconcha @MovGP0  @birbilis
I think validators are a good idea

**Comment by sethrudesill** [comment]
A standardized email address validator would be of huge benefit to users. Same for phone numbers and dates/times.

**Comment by insinfo** [comment]
@rdiazconcha @MovGP0 @birbilis @sethrudesill 
I think I could add text about Validators in this post or you could create a new post. I do not know how to do it.

If possible visit my posts to help in defining the model and syntax of these controls
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/98), 
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/99), 
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/109), 
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/110), 
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/112), 
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/117),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/120),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/121),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/122),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/123),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/124),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/125),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/126),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/127),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/128),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/129),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/130),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/131),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/132),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/135),
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/135),

**Comment by globeport** [comment]
Keep it as simple as possible in my opinion. I don't think XAML validators should be part of XAML standard. Validation should be in the model/business logic. MaxLength alters the behaviour of the control and should therefore remain as a standalone property.

**Comment by weitzhandler** [comment]
`IsReadOnly` should also be included.

**Comment by weitzhandler** [comment]
@globeport 
I'd love validation to be covered by the standard, but in a much wider way, and not tied to textbox. Maybe validation should be in a different namespace. Using attached properties etc.
validation operators shouldn't be here, but rather validation adorners and validation styling and reactions.

**Comment by globeport** [comment]
I agree, validation extensions would be helpful. My view is that generally people choose to present validation results in different ways. You might outline a textbox in red, or display a label or a popup. Anything that alters the behaviour of the control, e.g. MaxLength, IsReadOnly etc. should definitely be in the standard. But changing appearance or adding controls can easily be done with existing elements. Wrap this in an extension library but don't make it part of the standard.

**Comment by weitzhandler** [comment]
Any control should have a state where its style is set according to the state.

**Comment by MovGP0** [comment]
@sethrudesill technically, both email addresses and phone numbers are just URLs as defined by the following RFCs:

* [RFC 1123](https://tools.ietf.org/html/rfc1123) Requirements for Internet Hosts -- Application and Support
* [RFC 1738](https://tools.ietf.org/html/rfc1738)  Uniform Resource Locators (URL)
* [RFC 3492](https://tools.ietf.org/html/rfc3492) Punycode: A Bootstring encoding of Unicode for Internationalized Domain Names in Applications (IDNA)
* [RFC 3696](https://tools.ietf.org/html/rfc3696) Application Techniques for Checking and Transformation of Names
* [RFC 3966](https://tools.ietf.org/html/rfc3966)  The tel URI for Telephone Numbers
* [RFC 4291](https://tools.ietf.org/html/rfc4291) IP Version 6 Addressing Architecture
* [RFC 5321](https://tools.ietf.org/html/rfc5321) Simple Mail Transfer Protocol
* [RFC 5322](https://tools.ietf.org/html/rfc5322) Internet Message Format

(I might have forgotten something, but this pretty much covers it)

**Phone Number Parsing** 
I highly recommend using *libphonenumber* for phone number parsing: 

* JavaScript: [libphonenumber](https://github.com/googlei18n/libphonenumber)
* .NET: [libphonenumber-csharp](https://github.com/aidanbebbington/libphonenumber-csharp/tree/master) 

**Email Address Parsing**
* JavaScript: [email-addresses](https://github.com/jackbowman/email-addresses)
* .NET: wrap the URI class and check that `uri.Scheme == Uri.UriSchemeMailto`, as well as host name, username, empty password, etc.

**Comment by lthib** [comment]
_Replacing_ a `MaxLength` property is only appropriate if the replacement provides equivalent functionality. I wouldn't say that's the case with the suggestion in this thread. MaxLength necessarily constrains how many characters can be entered into a text box. A validator would allow arbitrarily long entries and then balk in some summary control somewhere that they've exceeded a particular count. The two both have merit but on separate accounts.

**Comment by weitzhandler** [comment]
Indeed, validation aside. MaxLength ReadOnly etc has nothing with validation, tho, would be nice to have integrated validation controlling these properties.
